psm/src/arch/loongarch64.s

```diff
@@ -0,0 +1,64 @@
+#include "psm.h"
+
+.text
+.globl rust_psm_stack_direction
+.align 2
+.type rust_psm_stack_direction,@function
+rust_psm_stack_direction:
+/* extern "C" fn() -> u8 */
+.cfi_startproc
+    li.w $r4, STACK_DIRECTION_DESCENDING
+    jr $r1
+.rust_psm_stack_direction_end:
+.size       rust_psm_stack_direction,.rust_psm_stack_direction_end-rust_psm_stack_direction
+.cfi_endproc
+
+
+.globl rust_psm_stack_pointer
+.align 2
+.type rust_psm_stack_pointer,@function
+rust_psm_stack_pointer:
+/* extern "C" fn() -> *mut u8 */
+.cfi_startproc
+    move $r4, $r3
+    jr $r1
+.rust_psm_stack_pointer_end:
+.size       rust_psm_stack_pointer,.rust_psm_stack_pointer_end-rust_psm_stack_pointer
+.cfi_endproc
+
+
+.globl rust_psm_replace_stack
+.align 2
+.type rust_psm_replace_stack,@function
+rust_psm_replace_stack:
+/* extern "C" fn(r4: usize, r5: extern "C" fn(usize), r6: *mut u8) */
+.cfi_startproc
+    move $r3, $r6
+    jr $r5
+.rust_psm_replace_stack_end:
+.size       rust_psm_replace_stack,.rust_psm_replace_stack_end-rust_psm_replace_stack
+.cfi_endproc
+
+
+.globl rust_psm_on_stack
+.align 2
+.type rust_psm_on_stack,@function
+rust_psm_on_stack:
+/* extern "C" fn(r4: usize, r5: usize, r6: extern "C" fn(usize, usize), r7: *mut u8) */
+.cfi_startproc
+    st.d $r1, $r7, -8
+    st.d $r3, $r7, -16
+    .cfi_def_cfa 7, 0
+    .cfi_offset 1, -8
+    .cfi_offset 3, -16
```

Comment:
Preexisting, given that this is probably mostly a port from the MIPS implementation, but I have a question… 

Are these actually necessary? The `r1` and `r3` don't appear to be actually clobbered until the `addi.d` instruction below, and the `.cfi_offset`s could easily be applied together with the `.cfi_def_cfa 3, -16`.

Either way this looks good to me correctness wise.

Reply:
> Preexisting, given that this is probably mostly a port from the MIPS implementation, but I have a question…

I actually used RISC-V implementation as a reference.

> Are these actually necessary? The `r1` and `r3` don't appear to be actually clobbered until the `addi.d` instruction below, and the `.cfi_offset`s could easily be applied together with the `.cfi_def_cfa 3, -16`.

I'll consider this and find some way to test (GDB currently does not work properly on LoongArch so I need to find another way).

Reply:
I've removed the unneeded `.cfi_def_cfa 7, 0` directive.  And `.cfi_def_cfa 3, -16` should be `.cfi_def_cfa 3, 16`, as `.cfi_def_cfa` is [defined](https://sourceware.org/binutils/docs/as/CFI-directives.html):

> 7.12.9 .cfi_def_cfa register, offset
>
> .cfi_def_cfa defines a rule for computing CFA as: take address from register and **add** offset to it.

I guess this also needs to be fixed for RISC-V (unless the RISC-V assembler has some special definition for CFI directives).

Tested using a C caller with glibc `backtrace()` function.